File: jcv-core/src/main/java/com/ekino/oss/jcv/core/validator/ValueParameterizedTemplateValidator.java
/*
 * Copyright (c) 2019 ekino (https://www.ekino.com/)
 */
package com.ekino.oss.jcv.core.validator;

import org.skyscreamer.jsonassert.ValueMatcher;

abstract class ValueParameterizedTemplateValidator<T> extends ValueTemplateIdValidator<T> {

    public ValueParameterizedTemplateValidator(String validatorId) {
        super(validatorId);
    }

    @Override
    public ValueMatcher<T> getValueComparator() {
        return (actual, expected) -> {
            if (expected instanceof String) {
                return getValueComparator(new ValidatorTemplateManager((String) expected)).equal(actual, expected);
            }
            throw new IllegalArgumentException("Invalid template definition : " + expected);
        };
    }

    protected abstract ValueMatcher<T> getValueComparator(ValidatorTemplateManager validatorTemplateManager);
}

File: jcv-core/src/main/kotlin/com/ekino/oss/jcv/core/validator/ValueParameterizedTemplateValidator.kt
/*
 * Copyright (c) 2019 ekino (https://www.ekino.com/)
 */
package com.ekino.oss.jcv.core.validator

import org.skyscreamer.jsonassert.ValueMatcher

abstract class ValueParameterizedTemplateValidator<T>(validatorId: String) : ValueTemplateIdValidator<T>(validatorId) {

    override val valueComparator: ValueMatcher<T>
        get() = ValueMatcher { actual, expected ->
            if (expected is String) {
                return@ValueMatcher getValueComparator(ValidatorTemplateManager(expected)).equal(actual, expected)
            }
            throw IllegalArgumentException("Invalid template definition : $expected")
        }

    protected abstract fun getValueComparator(validatorTemplateManager: ValidatorTemplateManager): ValueMatcher<T>
}
